### DISCUSSION TITLE
Google insights: We got unused CSS that we do not even import in page

### DISCUSSION THREAD

**Discussion by omoutop** [vercel/next.js#68184] [Help]
### Summary

We are building a project with NextJS 14 (App router) and we are facing issues with mobile speeds. One reason for bad performance scores is a lot of unused CSS on the page. But this CSS is not even imported into the page. It seems that NextJS is still creating batches with this CSS.

### Additional information

```js
Hi all!

We are building a project with NextJS 14 (App router) and we are facing issues with mobile speeds. One reason for bad performance scores is a lot of unused CSS on the page.

We have multiple CSS modules in our project. Almost every component has its own module.css file as per requirement.
We know that when the build is created, those CSS files are combined in NextJS, but the problem is that NextJS is combining CSS files that are not even required on our page.

Let's say we have Page A and Page B. Logically, Page A should have CSS files that have the styles of Page A only, right? But that is not the case. We see that in production, a CSS file is added to the head of Page A that contains many styles from Page B... We have double-checked that none of those styles are imported anywhere on Page A or the layout, but still, they are getting added and we can't figure out why.

We need a better understanding of how NextJS combines the CSS files which files are added to the document head and when they are added.

Thank you for your help as we are struggling to get more than 70+ in Google insights.
```

**Comment by icyJoseph** [comment]
Can you share a link to the coverage analysis?  

- https://nextjs.org/docs/app/building-your-application/styling/css-modules

> In production, all CSS Module files will be automatically concatenated into many minified and code-split .css files. These .css files represent hot execution paths in your application, ensuring the minimal amount of CSS is loaded for your application to paint.

With CSS Modules, and tailwind this is often a bit of an issue, but this is the first time that I see uncovered/unused CSS causing such drop. Are you sure this is the root issue of your performance insight? can you share a link so that I can take a look? There's often tree views of your modules, which also show the JS/CSS usage rate. ~Something like this: https://googlechrome.github.io/lighthouse/treemap/?gzip=1#~ Looks like that feature is broken now 😢 - but maybe paste a screenshot, or better yet a link to a pagespeed report.

**Reply by omoutop** [reply]
Hi, I don't think the drop is exclusively due to that. But we try to solve all our issues one by one.
Unfortunately, I am not allowed to share the URL publicly. 
I can send details on LinkedIn . Please Follow back.
